Use structured logging (metadata) in swift-rabbitmq

xtremekforever opened the pull request:
I briefly looking at using the `MetadataProvider`, but it seems like a `Logger` needs to be constructed using the provider and this can't be passed after the `Logger` exists. So instead we just copy the `logger` instance and set metadata needed for the current context.

 - Replaced `NullLogHandler` with `SwiftLogNoOpLogHandler`, which is included in swift-log.
 - Used structured logging throughout all sources, which results in output like this:
    ```
    info BasicConsumePublish : url=amqp://guest:guest@localhost/%2F [RabbitMq] Connecting to broker...
    info BasicConsumePublish : url=amqp://guest:guest@localhost/%2F [RabbitMq] Connected to broker
    trace BasicConsumePublish : exchangeName=MyTestExchange options=ExchangeOptions(type: RabbitMq.ExchangeType.direct, passive: false, durable: true, autoDelete: true, internal: false, args: [:]) [RabbitMq] Declaring exchange...
    trace BasicConsumePublish : options=QueueOptions(autoDelete: true, exclusive: false, durable: true, passive: false, args: [:]) queueName=MyTestQueue [RabbitMq] Declaring queue...
    ```

xtremekforever committed:
Use logger in BasicConsumePublish example

xtremekforever committed:
Add structured logging for all code

 - withRetryingConnectionBody still uses some string interpolation to make a readable string for the operation name, but apart from that everything else is structured.

xtremekforever committed:
Use SwiftLogNoOpLogHandler instead of NullLogHandler

 - They do the same thing, and the SwiftLogNoOpLogHandler is included with swift-log.

xtremekforever committed:
Add small test checks for logger metadata

xtremekforever committed:
Make AMQPChannel+Options metadata more clear

 - Use exchangeName and queueName instead of just "name", which can be confusing.

orobio commented on Sources/RabbitMq/Connection/RetryingConnection.swift:
```diff
@@ -23,8 +23,10 @@ public actor RetryingConnection: Connection, Service {
-    public let logger: Logger  // shared to users of Connection
     public let connectionPollingInterval: Duration
+    public var logger: Logger {
+        get async { await basicConnection.logger }
```
Doing an async call every time `logger` is used may not be ideal, but if it doesn't happen too often, I guess it's fine.

orobio commented on Sources/RabbitMq/Consumer/Consumer.swift:
```diff
@@ -16,10 +16,15 @@ import NIO
+
+    var logger: Logger {
+        get async {
+            await connection.logger.withMetadata(["queueName": .string(configuration.queueName)])
```
This creates a new logger with metadata every time `logger` is used. This is done even when the log is not actually created due to the current log level.

xtremekforever replied:
Good point- however, how can I ensure I always use the logger from connection that always contains the latest context (url)?

orobio commented on Sources/RabbitMq/Consumer/RetryingConsumer.swift:
```diff
@@ -3,24 +3,28 @@ import AsyncAlgorithms
 
+    var logger: Logger {
+        get async {
+            await connection.logger.withMetadata(["queueName": .string(configuration.queueName)])
```
This creates a new logger with metadata every time `logger` is used. This is done even when the log is not actually created due to the current log level.

orobio commented on Sources/RabbitMq/Publisher/Publisher.swift:
```diff
@@ -12,10 +12,15 @@ import NIOCore
+
+    var logger: Logger {
+        get async {
+            await connection.logger.withMetadata(["exchangeName": .string(configuration.exchangeName)])
```
This creates a new logger with metadata every time `logger` is used. This is done even when the log is not actually created due to the current log level.

orobio commented on Sources/RabbitMq/Publisher/RetryingPublisher.swift:
```diff
@@ -2,31 +2,38 @@ import AMQPClient
 
+    var logger: Logger {
+        get async {
+            await connection.logger.withMetadata(["exchangeName": .string(configuration.exchangeName)])
```
This creates a new logger with metadata every time `logger` is used. This is done even when the log is not actually created due to the current log level.

xtremekforever committed:
Merge branch 'main' into #25-structured-logging

xtremekforever committed:
Remove ConnectionLoggable and setting url metadata on BasicConnection

 - Instead we will just log with the context we have. Maybe there's a better solution out there for it.

xtremekforever committed:
Make retrying consumer/publisher sendable

xtremekforever committed:
Remove test expectations of metadata keys in loggers

xtremekforever committed:
Resolve async warnings around logger in Consumer and Publisher tests
